lib/ruby_indexer/lib/ruby_indexer/uri.rb

```diff
@@ -0,0 +1,108 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyIndexer
+  # The indexer Uri class is a specialization of the regular URI class. It includes convenience methods, special
+  # handling to support Windows paths and simplifications for URI elements that are not used in an LSP context.
+  #
+  # For example, it does not do anything with the `host` portion of the URI, but we include a `require_path` attribute,
+  # so that we know how this URI is mapped in the `$LOAD_PATH` if applicable
+  class Uri < URI::Generic
```

Comment:
Also, it looks like we always use `Uri.file` to create `FileUri` but never create `Uri` directly? Should we just have `FileUri` if that's the case?

Reply:
> I want to avoid calling it Uri as it's potentially confusing to readers who don't work with URI as often as we do. Can we call it something like IndexUri instead?

Given that all language server resources are always identified by URIs, I think we should maintain the standard and lean on the existing name. `IndexUri` is not accurate because the idea is for this URI object to be used everywhere in the language server and not just for indexing.

The idea would be to standardize all of our URI usage on that object.

> Also, it looks like we always use Uri.file to create FileUri but never create Uri directly? Should we just have FileUri if that's the case?

This is temporary. When we migrate the rest of the language server to use the new object, it will need to instantiate `Uri` because we receive requests for other schemes, like `git` and `untitled`.

Reply:
> IndexUri is not accurate because the idea is for this URI object to be used everywhere in the language server and not just for indexing.

But the definition originates from the indexer, which means it'd be used outside of language server in the future? I agree that `Index` prefix is a bit too restrictive so my point is just that we should ideally give it a clearer name.

Reply:
Settled on `ResourceUri`, which is generic enough to cover all scenarios, but not the same as `URI`.